lib/active_merchant/billing/gateways/omise.rb

```diff
@@ -0,0 +1,376 @@
+
+      def create_post_for_purchase(money, payment, options)
+        post = {}
+        post[:capture] = options[:capture] if options[:capture]
+        error_response = add_or_create_token(post, payment, options)
+        return error_response if error_response and error_response.is_a?(Response)
+        MultiResponse.run do |r|
+          r.process do
+            add_amount(post, money, options)
+            add_customer(post, options)
+            commit(:post, 'charges', post)
+          end
+        end.responses.last
+      end
+
+      def headers(options={})
+        key = options[:key] || @secret_key
+        {
+          "Content-Type"  => "application/json;utf-8",
+          "User-Agent"    => "Omise/v#{API_VERSION} ActiveMerchantBindings/#{ActiveMerchant::VERSION}",
+          'Authorization' => 'Basic ' + Base64.encode64(key.to_s + ':').strip,
+          'Accept-Encoding' => 'utf-8'
+        }
+      end
+
+      def post_data(params=nil)
+        return nil unless params
+        params.to_json
+      end
+
+      def https_request(method, endpoint, parameters=nil, options={})
+        raw_response = response = nil
+        url_endpoint = options[:url_endpoint] || API_URL + endpoint
+        begin
+          raw_response = ssl_request(method, url_endpoint, post_data(parameters), headers(options))
+          response = parse(raw_response)
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response = parse(raw_response)
+        rescue JSON::ParserError
+          response = json_error(raw_response)
+        end
+        response
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def json_error(raw_response)
+        {
+          message: 'Invalid response received from Omise API.  Please contact support@omise.co if you continue to receive this message.' +
+            '  (The raw response returned by the API was #{raw_response.inspect})'
+        }
+      end
+
+      def commit(method, endpoint, params=nil, options={})
+        response = https_request(method, endpoint, params, options)
+        Response.new(
+          successful?(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: successful?(response) ? nil : STANDARD_ERROR_CODE_MAPPING[error_code_from(response)]
+        )
+      end
+
+      def error_code_from(response)
+        (response['object'] == 'error') ? response['code'] : response['failure_code']
+      end
+
+      def message_from(response)
+        return 'Success' if successful?(response)
+        (response['message'] ? response['message'] : response['failure_message'])
+      end
+
+      def authorization_from(response)
+        response['id'] if successful?(response)
+      end
+
+      def successful?(response)
+        ( response.key?('object') and response['object'] != 'error' ) and response['failure_code'].nil?
```

Comment:
I see use of `and` and `or` a lot, which often doesn't do what you want. Instead use `&&` and `||` exclusively.